api/client/src/main/java/org/projectnessie/client/http/HttpResponse.java

```diff
@@ -34,49 +37,76 @@ public HttpResponse(ResponseContext context, ObjectMapper mapper) {
-        // but the tests in :nessie-versioned-gc-iceberg were running Nessie 0.4 server, which
-        // doesn't return a value, so `org.projectnessie.versioned.gc.TestUtils#resetData` failed.
-        return null;
-      }
-      try (InputStream is = responseContext.getInputStream()) {
-        if (!responseContext.isJsonCompatibleResponse()) {
-          nonJsonResponse();
-        }
+      return decodeEntity(reader, responseContext.getInputStream());
+    } catch (IOException e) {
+      throw new HttpClientException("Failed to read entity", e);
+    }
+  }
 
-        return reader.readValue(is);
+  private <V> V decodeEntity(ObjectReader reader, InputStream is) throws IOException {
+    if (is != null) {
+      CapturingInputStream capturing = new CapturingInputStream(is);
+      try {
+        if (responseContext.isJsonCompatibleResponse()) {
+          // readValues tolerates empty responses and returns an empty iterator in that case;
+          // this allows to handle these cases gracefully
+          MappingIterator<V> values = reader.readValues(capturing);
+          if (values.hasNextValue()) {
+            return values.nextValue();
+          }
+        } else {
+          String captured = capturing.capture();
+          if (!captured.isEmpty()) {
+            throw unparseableResponse(captured, null);
+          }
+        }
+      } catch (IOException e) {
+        throw unparseableResponse(capturing.capture(), e);
+      } finally {
+        capturing.close();
       }
-    } catch (IOException e) {
-      throw new HttpClientException("Cannot parse response.", e);
     }
+    // In case the Nessie server returns no content (e.g. because it's on an older version)
+    // but the client expects a response, just return null here and cross fingers that
+    // the code doesn't expect a non-null value.
+    return null;
   }
 
-  private void nonJsonResponse() throws IOException {
+  private NessieBadResponseException unparseableResponse(String captured, Exception cause) {
     Status status = responseContext.getStatus();
-    throw new NessieBadResponseException(
+    String message =
+        String.format(
+            "Expected the server to return a JSON compatible response, "
+                + "but the server replied with Content-Type '%s' from '%s' %s. "
+                + "Check the Nessie REST API base URI. "
+                + "Nessie REST API base URI usually ends in '/api/v1' or '/api/v2', "
+                + "but your service provider may have a different URL pattern.",
+            responseContext.getContentType(),
+            responseContext.getRequestedUri(),
+            captured.isEmpty()
+                ? "and no response body"
+                : "and a response body beginning with: \"" + captured + "\"");
+    NessieError error =
         ImmutableNessieError.builder()
+            .message(message)
             .status(status.getCode())
-            .message(status.getReason())
```

Comment:
I fixed the inconsistency here where the reason was being input as message, and the message as reason.

Reply:
Eagle eye ;)